### PR TITLE
feat: use clones-with-immutable-args

### PR DESCRIPTION
**Pull request by marktoda**
This commit uses the nice library ClonesWithImmutableArgs which allows
for minimal proxy clones to have immutable arguments. Currently they
cannot because immutable arguments must be created in the constructor.
This library allows them to be created by storing data in the contract's
code upon create-time, which can then be explicitly fetched using
calldataload, saving significant gas fees. In our case, ~3k gas for
deposit operations, ~2k gas for redemption operations, and ~7k gas for
contract creation

### CODE REVIEW

**Comment by marktoda** [comment]
https://github.com/wighawag/clones-with-immutable-args

File: contracts/BondController.sol [hunk 1]
```diff
@@ -27,47 +28,53 @@ contract BondController is IBondController, OwnableUpgradeable {
...
+    uint256 public override feeBps;
+
+    // Address of the collateral token
+    function collateralToken() public pure override returns (address) {
+        return _getArgAddress(0);
+    }
+
+    // Date at which this bond expires
+    function maturityDate() public pure override returns (uint256) {
+        return _getArgUint256(20);
```
[the remark this answers is not in the document]
Comment (marktoda) [reply contracts/BondController.sol:47]: always starts @0 and increments by arg size

[the remark this answers is not in the document]
Comment (marktoda) [reply contracts/BondController.sol:47]: ie address is 20 bytes

**Review by Fiddlekins — APPROVED** [review]

**Comment by aalavandhan** [comment]
my 2c here is that gains from gas optimization are close to insignificant at this point .. I'd rather we skip this and keep the code standard

**Comment by marktoda** [comment]
I think that's a fair assessment - 1-3% gas gains maybe not worth it. Worst case it was a cool idea :)

**Comment by marktoda** [comment]
Let me think on it a bit

**Comment by marktoda** [comment]
agree not worth it

**Pull request closed without merging**